security/pkg/nodeagent/secretfetcher/secretfetcher.go

```diff
@@ -443,6 +450,22 @@ func (sf *SecretFetcher) FindIngressGatewaySecret(key string) (secret model.Secr
 	val, exist := sf.secrets.Load(key)
 	secretFetcherLog.Debugf("load secret %s from secret fetcher: %v", key, exist)
 	if !exist {
+		// Sometimes we see that a secret in installed but not in cache because watcher is in an
+		// obsolete state and wasn't reset promptly. We bail this case out by trying fetching
+		// the secret from API call. Since this is a rare case, to avoid complication, we don't add
+		// the secret back to cache as it is not a normal codepath. When watcher recovers, those secret
+		// shall be added back.
```

Comment:
Perhaps mention in the comment that this approach only covers the TLS server key/cert fetching. I am thinking about other cases. For example, if key is "bookinfo-cacert", which is a k8s secret with ca cert. Then the return of extractK8sSecretIntoSecretItem() ignores ca secretItem and returns key/cert secretItem, which is always nil.

Reply:
In that case, because the returned key/cert secretItem is nil, the function will continue to its original path. We can check the return and return the ca secretItem if that is the right one to return, but what if both ca secretItem and key/cert SecretItem are non-nil?

Reply:
If user create a k8s secret called `bookinfo`, which has server key/cert and client CA cert, and configure gateway to be mTLS. Envoy will send two types of SDS requests, one request has resource name `bookinfo`, and citadel agent should return key/cert secretitem. The other has resource name `bookinfo-cacert`, and citadel agent should return ca secretitem. If extractK8sSecretIntoSecretItem() returns both secretitems that are non-nil, it depends on the request to decide which secretitem to return.

Reply:
FindIngressGatewaySecret() only have secret name as parameter, so I don't see how we can decide which one to return if they are both there. How about we return key/cert secretItem if it is not nil, otherwise, if isCAOnlySecret is true and ca secret item is non-nil, return the ca secretItem?

Reply:
That secret name passed to FindIngressGatewaySecret() is the SDS resouce name, not k8s secret name. The naming convention is a little misleading here. If the secret name (SDS resource name) is `bookinfo-cacert`, then it is asking for ca secretitem. This ca cert could belong to a compound k8s secret `bookinfo`, which has server key/cert and CA cert, or belong to a ca only k8s secret `bookinfo-cacert`, which only has CA cert. That SDS request for CA cert only happens when you set up a mTLS gateway instead of TLS gateway. 

The current approach in this PR keeps the behavior unchanged when Envoy asks for `bookinfo-cacert`.  You can leave a TODO here to cover mTLS gateway scenario.

Reply:
Thanks for the explanation. Added the comment for the limitation.